### PR TITLE
Group convolution

### PR DESCRIPTION
**Pull request by asleepzzz**
this PR enable group convolution,
padding and stride is ok
the format describe in mlir/docs/MIOpen/tensor_layout.md

we will test resnet101 f16 nhwc configs after f16 overflow issue fixed

### CODE REVIEW

**Review by whchung — CHANGES_REQUESTED** [review]
@asleepzzz / @ltqin In my first glance at the PR I don't see any e2e tests being added. Please help do that.

File: mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation_f16.mlir
Comment (whchung) [inline mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation_f16.mlir:128]: @asleepzzz this config is passing on `miopen-dialect`. please help confirm if it has become FLAKY again, or it's just a missed piece in the rebase process.

File: mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation.mlir
Comment (whchung) [inline mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation.mlir:141]: @asleepzzz this config is passing on miopen-dialect. please help confirm if it has become incorrect again, or it's just a missed piece in the rebase process.

File: mlir/test/Target/rocdl.mlir
Comment (whchung) [inline mlir/test/Target/rocdl.mlir:47]: @asleepzzz this test can *NOT* be removed.

File: mlir/test/Target/rocdl.mlir
Comment (whchung) [inline mlir/test/Target/rocdl.mlir]: @asleepzzz this check can *NOT* be removed.

File: mlir/test/Dialect/MIOpen/ops_2.mlir
Comment (whchung) [inline mlir/test/Dialect/MIOpen/ops_2.mlir:119]: @asleepzzz this test can *NOT* be removed.

File: mlir/test/Dialect/LLVMIR/rocdl.mlir
Comment (whchung) [inline mlir/test/Dialect/LLVMIR/rocdl.mlir:38]: @asleepzzz this test can *NOT* be removed.

File: mlir/test/Dialect/GPU/ops.mlir
Comment (whchung) [inline mlir/test/Dialect/GPU/ops.mlir:61]: @asleepzzz this test can *NOT* be removed.

File: mlir/test/Conversion/MIOpenToGPU/misc_ops.mlir
Comment (whchung) [inline mlir/test/Conversion/MIOpenToGPU/misc_ops.mlir:11]: @asleepzzz this test can *NOT* be removed.

File: mlir/test/Conversion/GPUToROCDL/gpu-to-rocdl.mlir
Comment (whchung) [inline mlir/test/Conversion/GPUToROCDL/gpu-to-rocdl.mlir:87]: @asleepzzz this test can *NOT* be removed.

**Review by whchung — CHANGES_REQUESTED** [review]
It seems all logic related to `miopen.lds_barrier` is somehow missing in this PR, as well as some test configs related with flakiness in fp16. Please help fix them.

**Comment by whchung** [comment]
@ltqin / @asleepzzz After the first iteration of going over the PR, I can think of 2 places be improved:
1. add e2e tests . It is crucial from now on to let CI help us continue monitor the correctness of the implementation.
2. fix lost information in the rebase process. `miopen.lds_barrier` and some critical flakiness checks were left out.

**Comment by whchung** [comment]
> this PR enable group convolution,
> padding and stride is ok
> the format describe in mlir/docs/MIOpen/tensor_layout.md
> 
> we will test resnet101 f16 nhwc configs after f16 overflow issue fixed

wondering have we tested resnext101 f32 nchw and nhwc cases with group > 1? I think they can be good candidates for e2e tests.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp
Comment (yiqian1) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: This should be %g = constant 103: i8. It's the ascii of 'g'.

File: mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp
Comment (yiqian1) [inline mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp]: This function is not needed anymore.

File: mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp
Comment (yiqian1) [inline mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:668]: This function is not needed because the internal representation would be always 5-dimensions, right?

**Comment by yiqian1** [comment]
And the two conv2d_harness_cpu_verifier* files under the e2e folder should be updated.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:514]: It may be a better idea to amend the comment to say we are emitting the ASCII codes for those letters.

**Comment by asleepzzz** [comment]
> @ltqin / @asleepzzz After the first iteration of going over the PR, I can think of 2 places be improved:
> 
> 1. add e2e tests . It is crucial from now on to let CI help us continue monitor the correctness of the implementation.
> 2. fix lost information in the rebase process. `miopen.lds_barrier` and some critical flakiness checks were left out.

2.done

**Comment by asleepzzz** [comment]
> @ltqin / @asleepzzz After the first iteration of going over the PR, I can think of 2 places be improved:
> 
> 1. add e2e tests . It is crucial from now on to let CI help us continue monitor the correctness of the implementation.
> 2. fix lost information in the rebase process. `miopen.lds_barrier` and some critical flakiness checks were left out.

2.done 
1. it's hard to add resnet101 configs, padding always >0 ,
for example
./bin/mlir-miopen-driver -p=false -fil_layout=gkcyx -in_layout=gnchw -out_layout=gnkhw -groupsize=32 -batchsize=256 -in_channels=1024 -out_channels=1024 -in_h=14 -in_w=14 -fil_h=3 -fil_w=3 --dilation_h=1 --dilation_w=1 --padding_h=1 --padding_w=1 --conv_stride_h=2 --conv_stride_w=2 -pr -c -ph | ./bin/mlir-rocm-runner --shared-libs=./lib/librocm-runtime-wrappers.so,./lib/libmlir_runner_utils.so --entry-point-result=void 
some results are
288<=in the center 192<= in the edge 128 <= corner

**Comment by asleepzzz** [comment]
resnet101 configs already added in 
mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation.mlir
mlir/test/mlir-miopen-driver/e2e/conv2d_host_validation_f16.mlir

fp32:
RESNET101  config1,config2
fwd ok, bwd,wrw crash , will fix it

config3-config7
couldn't find tunning parameters,
we need to find more tunning parameter

we can push this PR and fix resnet101 configs in the future

**Comment by asleepzzz** [comment]
driver can pass this test in  local but failed on ci
mlir-miopen-driver/cpu_conv2d/kyxc_nhwc_nhwk_bwd_data.mlir

mlir-miopen-driver -prc -fil_layout=kyxc -in_layout=nhwc -out_layout=nhwk  -batchsize=1 -in_channels=2 -out_channels=2 -in_h=6 -in_w=6 -fil_h=3 -fil_w=3 --dilation_h=2 --dilation_w=2 --padding_h=1 --padding_w=1 --conv_stride_h=2 --conv_stride_w=2 --operation=conv2d_bwd_data -p=false  -c | mlir-rocm-runner --shared-libs=/home/jenkins/workspace/MLIR/MLIR-downstream-unit-test/build/lib/librocm-runtime-wrappers.so,/home/jenkins/workspace/MLIR/MLIR-downstream-unit-test/build/lib/libmlir_runner_utils.so --entry-point-result=void | /home/jenkins/workspace/MLIR/MLIR-downstream-unit-test/build/bin/FileCheck /home/jenkins/workspace/MLIR/MLIR-downstream-unit-test/mlir/test/mlir-miopen-driver/cpu_conv2d/kyxc_nhwc_nhwk_bwd_data.mlir --check-prefix=CHECK16

**Comment by jerryyin** [comment]
@whchung Can we hold off merging this PR until I have a MIOpen code review addressed? Once this PR is in, I will be forced to amend all existing MIOpen PRs since it changed the device interface.

By tomorrow I should be able to have a interface PR that updated the `Miir.h`. My ask is to merge this PR at least after that is done. I will leave another comment when it is ready to merge from my perspective.

@sjw36 for awareness.

**Comment by whchung** [comment]
> @whchung Can we hold off merging this PR until I have a MIOpen code review addressed? Once this PR is in, I will be forced to amend all existing MIOpen PRs since it changed the device interface.

@jerryyin This PR is really big, and we are starting to cover some configs of resnext101. The more we wait the more rebase / cherry-pick @asleepzzz have to do and this holds off their next assignment, bwd v4r1.

I plan to take the PR after:
- @asleepzzz and @ltqin provide us a spreadsheet covering what's supported, and what's NOT supported in resnext101 configs.
- along with a list of remaining tasks to be done.

We shall address your concern in the weekly meeting today.

**Comment by asleepzzz** [comment]
[resnet101_group_configs.xlsx](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/files/6312857/resnet101_group_configs.xlsx)

**Pull request closed without merging**
